config/machine_id.go

```diff
@@ -0,0 +1,152 @@
+package config
+
+import (
+	"crypto/rand"
+	"fmt"
+	"log"
+	"strings"
+
+	"github.com/NordSecurity/nordvpn-linux/internal"
+	"github.com/google/uuid"
+)
+
+type MachineIDGetter interface {
+	GetMachineID() uuid.UUID
+}
+
+type FileReader func(fileName string) ([]byte, error)
+type HostNameReader func() (name string, err error)
+
+type GeneratorFn func() ([]byte, error)
+
+type MachineID struct {
+	hostNameReader HostNameReader
+	fileReader     FileReader
+}
+
+func NewMachineID(fileReader FileReader, hostNameReader HostNameReader) MachineID {
+	return MachineID{
+		hostNameReader: hostNameReader,
+		fileReader:     fileReader,
+	}
+}
+
+func (getter MachineID) GetMachineID() uuid.UUID {
+	id, err := getter.generateID()
+	if err == nil {
+		return id
+	}
+
+	// create a random UUID
+	log.Println(internal.ErrorPrefix, "failed to generate machine ID", err)
+	id, err = uuid.NewRandom()
+	if err == nil {
+		return id
+	}
+
+	// Fallback to manually generating a UUID
+	log.Println(internal.ErrorPrefix, "failed to generate random UUID", err)
+	var fallbackUUID uuid.UUID
+	_, err = rand.Read(fallbackUUID[:])
+	if err != nil {
+		log.Println(internal.ErrorPrefix, "rand failed, retry to generate uuid", err)
+		return uuid.New()
+	}
+
+	// Set version (4) and variant bits according to RFC 4122
+	fallbackUUID[6] = (fallbackUUID[6] & 0x0F) | 0x40 // Version 4
+	fallbackUUID[8] = (fallbackUUID[8] & 0x3F) | 0x80 // Variant (10xx)
+
+	// ensure that the random generate UUID is correct
+	return uuid.MustParse(fallbackUUID.String())
+}
+
+func (getter MachineID) generateID() (uuid.UUID, error) {
+	machineId, err := getter.calculateMachineID()
+	if err != nil {
+		return machineId, fmt.Errorf("failed to generate machine ID %w", err)
+	}
+
+	// order is important, changing the order would create new machine ID
+	generators := []GeneratorFn{
+		getter.readCPUSerial,
+		getter.readProductUUID,
+		getter.readMotherboardSerialNumber,
+	}
+
+	for _, generatorFn := range generators {
+		if val, err := generatorFn(); err == nil {
+			machineId = uuid.NewSHA1(machineId, val)
+		}
+	}
+
+	return machineId, nil
+}
+
+// get the machine UUID from one of the files and combine it with the hostname
+func (getter MachineID) calculateMachineID() (uuid.UUID, error) {
+	sourceFiles := []string{
+		"/etc/machine-id",
+		"/var/lib/dbus/machine-id",
+		"/sys/class/dmi/id/product_uuid",
+	}
+
+	hostname, err := getter.hostNameReader()
+	if err != nil {
+		return uuid.UUID{}, fmt.Errorf("failed to read hostname %w", err)
```

Comment:
Maybe `nil` should be here return 1st param

Reply:
cannot return nil, unless the function return type is changed to `*uuid.UUID`